streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -252,6 +288,111 @@ public void buildAndOptimizeTopology() {
         }
     }
 
+    private void maybePerformOptimizations(final Properties props) {
+
+        if (props != null && props.getProperty(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE)) {
+            LOG.debug("Optimizing the Kafka Streams graph for repartition nodes");
+            maybeOptimizeRepartitionOperations();
+        }
+    }
+
+    private void maybeOptimizeRepartitionOperations() {
+        StreamsGraphNode optimizedSingleRepartition = null;
+
+        for (final Map.Entry<StreamsGraphNode, Set<OptimizableRepartitionNode>> streamsGraphNodeSetEntry : keyChangingOperationsToOptimizableRepartitionNodes.entrySet()) {
+
+            StreamsGraphNode keyChangingNode = streamsGraphNodeSetEntry.getKey();
+
+            if (streamsGraphNodeSetEntry.getValue().isEmpty()) {
+                continue;
+            }
+
+            final StreamSourceNode parentSourceNode = (StreamSourceNode) findParentNodeMatching(keyChangingNode, n -> n instanceof StreamSourceNode);
+
+            if (parentSourceNode == null) {
+                throw new IllegalStateException(String.format("Can't find parent source node for %s ", keyChangingNode));
+            }
+
+            optimizedSingleRepartition = createRepartitionNode(keyChangingNode.nodeName(),
+                                                               parentSourceNode.keySerde(),
+                                                               parentSourceNode.valueSerde());
+
+            // re-use parent id to make sure the single repartition graph node is evaluated before downstream nodes
+            optimizedSingleRepartition.setId(parentSourceNode.id());
+
+            for (final OptimizableRepartitionNode repartitionNodeToBeReplaced : streamsGraphNodeSetEntry.getValue()) {
+
+                final StreamsGraphNode keyChangingNodeChild = findParentNodeMatching(repartitionNodeToBeReplaced, gn -> gn.parentNodes().contains(keyChangingNode));
```

Comment:
This loop only needs to happen once, so it seems like you can move it outsize the enclosing loop.

Reply:
This loop needs to remain where it is.  A single key-changing node could spawn multiple repartition nodes later in the graph.  The repartition nodes that are replaced are grouped by the common key-changing graph node.  The included integration test `RepartitionOptimizingIntegrationTest` contains an example of this situation.